### PR TITLE
Feature/extarnal database

### PR DESCRIPTION
**Pull request by mbarbet**

### CODE REVIEW

File: src/app/app.module.ts [hunk 1]
```diff
@@ -147,6 +148,22 @@ export function configUpdaterFactory(x): any {
   return x[0];
 }
 
+export function getOptionsFactory(arlasAuthService: AuthentificationService): any {
+  const getOptions = () => {
+    const token = arlasAuthService.accessToken !== null ? arlasAuthService.accessToken : null;
```
Comment (laurent-thiebaud-gisaia) [inline src/app/app.module.ts]: No sure you need the test (if it is null, then you manually set null)

File: src/app/app.module.ts [hunk 2]
```diff
@@ -147,6 +148,22 @@ export function configUpdaterFactory(x): any {
   return x[0];
 }
 
+export function getOptionsFactory(arlasAuthService: AuthentificationService): any {
+  const getOptions = () => {
```
Comment (laurent-thiebaud-gisaia) [inline src/app/app.module.ts]: No sure you need the getOptions variable, maybe this is expected (you can also `return () => {...`

File: src/app/components/bookmark-menu/bookmark-menu.component.ts [hunk 3]
```diff
@@ -45,12 +46,22 @@ export class BookmarkMenuComponent implements OnInit {
...
-    this.bookmarkService.dataBase.dataChange.subscribe(bookmarks => {
-      const sortedBookmark = bookmarks.sort((a, b) => {
-        return (a.views < b.views ? -1 : 1) * (-1);
-      });
-      this.topBookmarks = sortedBookmark.slice(0, this.nbTopBookmarks);
-    });
+    if (this.bookmarkService.dataBase instanceof BookmarkPersistenceDatabase) {
+      (this.bookmarkService.dataBase as BookmarkPersistenceDatabase).dataChange
+        .subscribe((bookmarks: { total: number, items: BookMark[] }) => {
+          const sortedBookmark = bookmarks.items.sort((a, b) => {
```
Comment (laurent-thiebaud-gisaia) [inline src/app/components/bookmark-menu/bookmark-menu.component.ts]: There seems to be the same code in both cases `const sortedBookmark = ...`

File: src/app/services/extend/extend.service.spec.ts [hunk 4]
```diff
@@ -17,34 +17,58 @@
...
-      ArlasStartupService, HttpClient, ArlasConfigService,
-      ArlasCollaborativesearchService, TranslateService, TranslateStore,
-      { provide: CONFIG_UPDATER, useValue: {} }
-    ]
-  }));
 
-  it('should be created', () => {
-    const service: ArlasExtendService = TestBed.get(ArlasExtendService);
-    expect(service).toBeTruthy();
+  // beforeEach(() => TestBed.configureTestingModule({
```
Comment (laurent-thiebaud-gisaia) [inline src/app/services/extend/extend.service.spec.ts]: Can it be deleted, or still required for later?

File: src/app/services/extend/extend.service.ts [hunk 5]
```diff
@@ -18,33 +18,68 @@
...
-import { ArlasStartupService } from '../startup/startup.service';
+import { HttpClient } from '@angular/common/http';
 
-@Injectable({
-  providedIn: 'root'
-})
+@Injectable()
 export class ArlasExtendService {
-  public dataBase: ExtendDatabase;
+  public dataBase: ExtendLocalDatabase | ExtendPersistenceDatabase;
```
Comment (laurent-thiebaud-gisaia) [inline src/app/services/extend/extend.service.ts]: Is there no parent class or interface to both ExtendLocalDatabase & ExtendPersistenceDatabase?

File: src/app/app.module.ts [hunk 6]
```diff
@@ -147,6 +148,22 @@ export function configUpdaterFactory(x): any {
...
+    const token = arlasAuthService.accessToken !== null ? arlasAuthService.accessToken : null;
+    if (token !== null) {
```
Comment (MohamedHamouGisaia) [inline src/app/app.module.ts]: I think we should test `undefined` too : `token !== null && token !== undefined`

File: src/config.json [hunk 7]
```diff
@@ -958,9 +958,9 @@
       }
     },
     "server": {
-      "url": "http://localhost:9999/arlas",
+      "url": "http://demo.arlas.io/arlas",
```
Comment (MohamedHamouGisaia) [inline src/config.json]: should we keep `http://localhost:9999/arlas` and `geodata` ?

File: src/app/app.module.ts (hunk 2 above)
Comment (sebbousquet) [reply src/app/app.module.ts]: The getOptions variable is needed because of the compiler. 
He doesn't support returning a lambda function : 
https://angular.io/guide/aot-compiler#no-arrow-functions

The compiler is ok with assigning the lambda function to a variable and return it.

File: src/app/components/bookmark-menu/bookmark-menu.component.ts (hunk 3 above)
Comment (sebbousquet) [reply src/app/components/bookmark-menu/bookmark-menu.component.ts]: It's the iterator which is different

File: src/app/services/extend/extend.service.spec.ts (hunk 4 above)
Comment (sebbousquet) [reply src/app/services/extend/extend.service.spec.ts]: Required for later

File: src/app/services/extend/extend.service.ts (hunk 5 above)
Comment (sebbousquet) [reply src/app/services/extend/extend.service.ts]: There is no common thing that could be used here.

File: src/app/app.module.ts (hunk 6 above)
Comment (sebbousquet) [reply src/app/app.module.ts]: Fixed

File: src/config.json (hunk 7 above)
Comment (sebbousquet) [reply src/config.json]: Fixed

File: src/app/app.module.ts (hunk 1 above)
Comment (sebbousquet) [reply src/app/app.module.ts]: Use of `!!` to also test undefined
